### ISSUE TITLE
Setting instance weights for ranking tasks

### ISSUE THREAD

**Issue by linkin8834** [dmlc/xgboost#4168]
Hi there,

I want to set up weights on two different levels for my *ranking* task:
* one is on the item-level (say, I want to set different weights to different items inside the *same* query group)
* The other is on the query level (say, I want to set different weight for different query groups)

I am going to use the binary version `xgboost` (not Python, Scala, etc.) for training and I am using `libsvm` formatted training data. Is it possible to do the aforementioned two tasks in this scenario? 

Thanks!

**Comment by hcho3** [comment]
Currently, you may only set weights on the query group level. If your ranking metric is NDCG, you can use different relevance judgment label (1, 2, 3, 4, etc) to differentiate individual items within the same group.

**Comment by linkin8834** [comment]
@hcho3 Gotya! 

*set weights on query group level*: is it possible to do that by using the binary xgboost, or it is only available for Python?

**Comment by hcho3** [comment]
You should create a file named `XXX.libsvm.weight` where `XXX.libsvm` is the input file.

**Comment by linkin8834** [comment]
@hcho3 Thanks!!!!

**Comment by linkin8834** [comment]
@hcho3 I am wondering how should I set the query-group-level instance weight in a *multi-node distributed training scenario*? The format in https://xgboost.readthedocs.io/en/latest/tutorials/input_format.html#instance-weights seems to be designed for setting the instance weight on the `single instance level`.

**Comment by hcho3** [comment]
Unfortunately, I don’t think it is currently possible. We need to come up with a better way.

**Comment by linkin8834** [comment]
@hcho3 thanks for the quick reply!!!
